File: basiclib/private/external/demo/Merger.java
/* Example of using BETA-like Component.java to
   implement a coroutine package for Java.

   class Coroutine is a superclass for Java coroutines

*/

class Adder extends Coroutine {
    public int res;
    int start, max;
    public Adder(int s, int m) { 
      start = s; max = m; 
    }
    void compute(int V){
	res = V+V;
        suspend();
	if (V < max) compute(V+1);
    }
    public void Do() {
	compute(start);
	suspend();
	res = 1000000;
    }
}
class Multiplier extends Coroutine {
    public int res;
    int start, max;
    public Multiplier(int s, int m) { 
      start = s; max = m; 
    }
    void compute(int V){
	res = V*V;
        suspend();
	if (V < max) compute(V+1);
    }
    public void Do() {
	compute(start);
	suspend();
	res = 1000000;
    }
}

class Merger {
    Adder P1 = new Adder(3,6);
    Multiplier P2 = new Multiplier(2,5);
    void merge() {
	int i;
        P1.call();
        P2.call();
      	for (i=0; i<6; i++){
	    if (P1.res < P2.res) {
		System.out.println("P1: " + P1.res);
		P1.call();
	    }
	    else {
		System.out.println("P2: " + P2.res);
		P2.call();
	    }
	}
    }      
    public static void main(String args[]) {
	Merger merger = new Merger();
	merger.merge();
    }
}

File: compiler/DOC/BYTECODE/coroutines/csharp/Merger.cs
/* Example of using BETA-like Component.cs to
   implement a coroutine package for C#.

   class Coroutine is a superclass for C# coroutines
 */

using System;

class Adder: Coroutine {
    public int res;
    int start;
    public Adder(int s) { 
      start = s; 
    }
    void compute(int V){
	res = V+V;
        A: suspend();
	compute(V+1);
    }
    public override void Do() {
	compute(start);
    }
}
class Multiplier: Coroutine {
    public int res;
    int start;
    public Multiplier(int s) { 
      start = s;
    }
    void compute(int V){
	res = V*V;
        M: suspend();
	compute(V+1);
    }
    public override void Do() {
	compute(start);
	suspend();
    }
}

class Merger: Coroutine {
    Adder A = new Adder(3);
    Multiplier M = new Multiplier(2);
    public override void Do() {
        A.call();
        M.call();
      	for (int i=0; i<6; i++){
	    if (A.res < M.res) {
		System.Console.WriteLine("A: " + A.res);
		A.call();
	    } else {
		System.Console.WriteLine("M: " + M.res);
		M.call();
	    }
	}
    }      
    public static void Main(String[] args) {
	Merger merger = new Merger();
	merger.call();
    }
}

